pkg/querier/batch/batch.go

```diff
@@ -129,25 +129,39 @@ func (a *iteratorAdapter) Next() chunkenc.ValueType {
 	}
 
 	if a.curr.Index < a.curr.Length {
-		// TODO native histograms: return correct batch type
-		return chunkenc.ValFloat
+		return a.curr.ValueType
 	}
 	return chunkenc.ValNone
 }
 
 // At implements chunkenc.Iterator.
 func (a *iteratorAdapter) At() (int64, float64) {
+	if a.curr.ValueType != chunkenc.ValFloat {
+		panic(fmt.Sprintf("Cannot read float from batch %v", a.curr.ValueType))
+	}
 	return a.curr.Timestamps[a.curr.Index], a.curr.Values[a.curr.Index]
 }
 
 // AtHistogram implements chunkenc.Iterator.
 func (a *iteratorAdapter) AtHistogram() (int64, *histogram.Histogram) {
-	panic("iteratorAdapter.AtHistogram not yet implemented")
+	if a.curr.ValueType != chunkenc.ValHistogram {
+		panic(fmt.Sprintf("Cannot read histogram from batch %v", a.curr.ValueType))
+	}
+	return a.curr.Timestamps[a.curr.Index], (*histogram.Histogram)(a.curr.PointerValues[a.curr.Index])
 }
 
 // AtFloatHistogram implements chunkenc.Iterator.
 func (a *iteratorAdapter) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
-	panic("iteratorAdapter.AtFloatHistogram not yet implemented")
+	// The promQL engine works on Float Histograms even if the underlying data is an integer histogram
+	// and will call AtFloatHistogram on a Histogram
+	if a.curr.ValueType == chunkenc.ValFloatHistogram {
+		return a.curr.Timestamps[a.curr.Index], (*histogram.FloatHistogram)(a.curr.PointerValues[a.curr.Index])
+	}
+	if a.curr.ValueType == chunkenc.ValHistogram {
+		h := (*histogram.Histogram)(a.curr.PointerValues[a.curr.Index])
+		return a.curr.Timestamps[a.curr.Index], h.ToFloat()
```

Comment:
I'm going to refer to @beorn7  on this one, because:

https://github.com/prometheus/prometheus/blob/5ec1b4baaf01dae0b22567db53886c83b74e0e76/promql/engine.go#L1779  
https://github.com/prometheus/prometheus/blob/5ec1b4baaf01dae0b22567db53886c83b74e0e76/promql/engine.go#L1894
https://github.com/prometheus/prometheus/blob/5ec1b4baaf01dae0b22567db53886c83b74e0e76/promql/engine.go#L1924

I will note however that in this experimental phase, we aim for correctness, not optimizations for histograms, so I'd defer this to later. Open an issue?

Reply:
I've double checked and our integration tests fail (on sparsehistogram branch) if I remove this automatic upconvert

Reply:
If the histogram was an integer histogram, and if the caller asks for `ToFloat()`, then we do need to return the converted float histogram. The conversion ends up requiring deep copy because it's a different struct and type of slice gets changed (int64 to float). The only thing common are the histogram spans, which are small compared to the bucket slices in general so we can just copy away those spans for simplicity.

Reply:
For context to Marco: The difference between integer and float histogram is that integer histogram stores buckets with delta encoding, while float histograms have _absolute_ float values for buckets, which is required by the query engine to work on.

Reply:
Thanks for answering, @codesome .

Note that, in Prometheus, by calling `it.AtFloatHistogram()`, we avoid the additional copy because the implementation of the `AtFloatHistogram` method will create a `FloatHistogram` directly from the encoded form of an integer histogram (rather than first materializing an integer `Histogram` and then calling `ToFloat` on it).

Relevant code in https://github.com/prometheus/prometheus/blob/main/tsdb/chunkenc/histogram.go , in particular https://github.com/prometheus/prometheus/blob/d33eb3ab17616a54b97d9f7791c791a79823f279/tsdb/chunkenc/histogram.go#L684-L701